src/typeahead/typeahead.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  Directive,
+  OnInit,
+  Input,
+  ComponentRef,
+  ComponentFactoryResolver,
+  ViewContainerRef,
+  Injector,
+  Renderer,
+  ElementRef
+} from '@angular/core';
+import {Observable} from 'rxjs/Rx';
+import 'rxjs/add/operator/filter';
+import 'rxjs/add/operator/let';
+import {Positioning} from '../util/positioning';
+import {NgbTypeaheadWindow} from './typeahead-window';
+import {PopupService} from '../util/popup';
+import {NgModel} from '@angular/common';
+import {NgbHighlight} from './highlight';
+
+// TODO: prevent default keyup behavior
+@Directive({
+  selector: 'input[ngbTypeahead][ngModel]',
+  host: {
+    '[class.open]': '_windowRef !== null',
+    '(document:click)': '_closePopup()',
+    '(keyup.arrowDown)': '_onKeyDown()',
+    '(keyup.arrowUp)': '_onKeyUp()',
+    '(keyup.enter)': '_onKeyEnter()',
+    '(keyup.esc)': '_closePopup()'
+  }
+})
+export class NgbTypeahead implements OnInit {
+  private _popupService: PopupService<NgbTypeaheadWindow>;
+  private _positioning = new Positioning();
+  private _skipValueChange = false;
+  private _windowRef: ComponentRef<NgbTypeaheadWindow>;
+
+  @Input() ngbTypeahead: (text: Observable<string>) => Observable<string[]>;
```

Comment:
It will also need an input for a `formatter` (so we can pass it down to the typeahead window).